chain/chain/src/store.rs

```diff
@@ -2933,6 +2933,78 @@ impl<'a> ChainStoreUpdate<'a> {
 
         Ok(())
     }
+
+    pub fn save(self) -> SavedStoreUpdate {
+        SavedStoreUpdate {
+            store_updates: self.store_updates,
+            chain_store_cache_update: self.chain_store_cache_update,
+            head: self.head,
+            tail: self.tail,
+            chunk_tail: self.chunk_tail,
+            fork_tail: self.fork_tail,
+            header_head: self.header_head,
+            final_head: self.final_head,
+            largest_target_height: self.largest_target_height,
+            trie_changes: self.trie_changes,
+            add_state_changes_for_split_states: self.add_state_changes_for_split_states,
+            remove_state_changes_for_split_states: self.remove_state_changes_for_split_states,
+            add_blocks_to_catchup: self.add_blocks_to_catchup,
+            remove_blocks_to_catchup: self.remove_blocks_to_catchup,
+            remove_prev_blocks_to_catchup: self.remove_prev_blocks_to_catchup,
+            add_state_dl_infos: self.add_state_dl_infos,
+            remove_state_dl_infos: self.remove_state_dl_infos,
+            challenged_blocks: self.challenged_blocks,
+        }
+    }
+}
+
+/// Saves changes from ChainStoreUpdate without link to ChainStore. Needed to preserve changes while
+/// applying block in other thread
+pub struct SavedStoreUpdate {
```

Comment:
Couple of alternatives to avoid copy-paste:

a) Change `ChainStoreUpdate` to be 

```
ChainStoreUpdate<'a> {
  chain_store: &'a mut ChainStore,
  data: SavedStoreUpdate // but with some different name
}
```

b) change it to be 

```
type ChainStoreUpdate<'a>  = ChainStoreUpdateImpl<&'a mut ChainStore> ;
type SavedChainStoreUpdate<'a>  = ChainStoreUpdateImpl<()> ;
pub struct ChainStoreUpdateImpl<T> {
    chain_store: T,
    store_updates: Vec<StoreUpdate>,
```

Reply:
b look nice, will try to do it